### ISSUE TITLE
Queue not kept alive on PyPy, and greenlet doing q.get() disappears

### ISSUE THREAD

**Issue by antocuni** [gevent/gevent#1519]
* gevent version: git revision 36704768
* Python version: 2.7.16 (Ubuntu 19.04), PyPy2.7 7.3.0 (downloaded from pypy.org)
* Operating System: `Ubuntu 19.04 5.0.0-38-generic #41-Ubuntu SMP Tue Dec 3 00:27:35 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

Consider this python file, which implements a classical producer/consumer scheme. 
```python
import sys
print sys.version
import gevent
gevent.config.loop = 'libev-cffi'
import gevent.core
import gevent.queue
print 'gevent version:', gevent.__version__
print 'gevent loop:', gevent.core.loop
print

import itertools
import greenlet
import gc
from weakref import WeakSet

class MyTest(object):

	def __init__(self):
		self.queues = WeakSet()

	def publish(self, item):
		for q in self.queues:
			q.put_nowait(item)

	def consumer(self):
		q = gevent.queue.Queue()
		self.queues.add(q)
		for i in itertools.count():
			print 'getter %d: q.get()' % i
			item = q.get()
			print 'getter %d: got item %s' % (i, item)

	def producer(self):
		for i in range(3):
			gc.collect()
			print 'Publishing %d (%d queues alive)' % (i, len(list(self.queues)))
			self.publish('hello world %d' % i)
			gevent.sleep()

def main():
	mytest = MyTest()
	gevent.spawn(mytest.consumer)
	g = gevent.spawn(mytest.producer)
	g.join()
	print 'bye'

if __name__ == '__main__':
	main()

```

Let's look at the output on PyPy and CPython:
```
$ PURE_PYTHON=1 python producer_consumer.py 
2.7.16 (default, Oct  7 2019, 17:36:04) 
[GCC 8.3.0]
gevent version: 1.5a4.dev0
gevent loop: <class 'gevent.libev.corecffi.loop'>

getter 0: q.get()
Publishing 0 (1 queues alive)
getter 0: got item hello world 0
getter 1: q.get()
Publishing 1 (1 queues alive)
getter 1: got item hello world 1
getter 2: q.get()
Publishing 2 (1 queues alive)
getter 2: got item hello world 2
getter 3: q.get()
bye

$ PURE_PYTHON=1 pypy producer_consumer.py 
2.7.13 (724f1a7d62e8, Dec 23 2019, 15:36:24)
[PyPy 7.3.0 with GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
gevent version: 1.5a4.dev0
gevent loop: <class 'gevent.libev.corecffi.loop'>

getter 0: q.get()
Publishing 0 (0 queues alive)
Publishing 1 (0 queues alive)
Publishing 2 (0 queues alive)
bye
```

Some things to note:
- I am using `libef-cffi` as the loop, and running it with `PURE_PYTHON=1`, to make sure to execute the same code on PyPy and CPython
- the only reference to the `Queue` object created inside `consumer` is the `q` local variable
- the result of `gevent.spawn(mytest.consumer)` is NOT saved, so there is no explicit reference to the greenlet anywhere

On CPython, it seems to work; on PyPy, after I do the `gc.collect()`, the `Queue` object and the `consumer` greenlet seems to be garbage-collected by the GC, so `consumer` is never waked up.
If you save either the queue or the greenlet somewhere, the GC does not collect them and the code works as expected.

I managed to write an even smaller reproducer:
```python
import sys
import weakref
import gc
print sys.version

import gevent
gevent.config.loop = 'libev-cffi'
import gevent.core
import gevent.queue
print 'gevent version:', gevent.__version__
print 'gevent loop:', gevent.core.loop
print

WQ = None
def producer():
	global WQ
	q = gevent.queue.Queue()
	WQ = weakref.ref(q)
	print 'q.get()'
	q.get()

def main():
	gevent.spawn(producer)
	gevent.sleep()
	print 'before GC:', WQ()
	gc.collect()
	print 'after GC:', WQ()
	import objgraph
	objgraph.show_backrefs(WQ(), refcounts=True)

main()
```

```
$ PURE_PYTHON=1 python weakqueue.py 
2.7.16 (default, Oct  7 2019, 17:36:04) 
[GCC 8.3.0]
gevent version: 1.5a4.dev0
gevent loop: <class 'gevent.libev.corecffi.loop'>

q.get()
before GC: <Queue getters[1]>
after GC: <Queue getters[1]>

$ PURE_PYTHON=1 pypy weakqueue.py 
2.7.13 (724f1a7d62e8, Dec 23 2019, 15:36:24)
[PyPy 7.3.0 with GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
gevent version: 1.5a4.dev0
gevent loop: <class 'gevent.libev.corecffi.loop'>

q.get()
before GC: <Queue getters[1]>
after GC: None
```

Here it is more clear that on PyPy, nobody keeps alive the `Queue`, which in turns is supposed to keep alive the greenlet who is blocked on its `.get()`.

I tried to dig inside the gevent code, but I cannot understand **who** is supposed to keep alive the queue, even on CPython. The folloing is a screenshot of the output of `objgraph` on CPython: you can see that the `Queue` object has a refcount of 11, but only 4 incoming arrows. Where do the extra 7 incref comes from?

![image](https://user-images.githubusercontent.com/344507/73123452-5dd06380-3f90-11ea-8c42-0ff354ee6ffc.png)

Unfortunately, I can't go deeper than that because I don't know much of `gevent` internals. It might be a PyPy issue, so I'd like some input from `gevent` devs on that.

**Comment by jamadden** [comment]
Interesting. (Comments below refer to the first version of the example program.)

Most often, in gevent, waiting greenlets are kept alive simply because they're waiting. That is, a blocking API like `socket.recv` results in the a reference to `getcurrent().switch` being held by the gevent hub (often through a couple layers of indirection —most waitable objects have a list of callables they call when they become ready; this is usually the abstraction referred to as "the link protocol").

That's not quite how a Queue works. In the case of a Queue, the `get()` method creates an instance of the `Waiter` class as a local variable, and stores it in a list instance variable. The `Waiter` stores the current greenlet in an instance variable, and then switches to the hub. When an object is put in the queue, a callback (queue instance method) is scheduled with the hub. When the callback runs, a `Waiter` is taken off the ivar list and its greenlet switched to. 

Up until the callback is scheduled, there's no place where the hub would be keeping such a Queue or the consumer greenlet alive. The references to it are contained within the Queue object itself and the stack. And that appears to be exactly what PyPy's GC thinks too: As soon as `q.get()` is executed and the hub is switched to, the stack becomes non-current and the Queue and greenlet are identified as a sub-graph of objects unreachable from anywhere else and are collected by the producer's `gc.collect()`.  

An interesting clue comes from the fact that after toying around a bit, I'm not able to get CPython to recognize that consumer greenlet or Queue as unreferenced *at all*. It never vanishes. So I suspect we are hitting this clause of the CPython greenlet documentation:

> Greenlets do not participate in garbage collection; cycles involving data that is present in a greenlet’s frames will not be detected. Storing references to other greenlets cyclically may lead to leaks.

This would be the cycle: getcurrent() -> `consumer` frame -> Queue -> Waiter -> getcurrent(). The same Waiter is referenced from the stack

This same cycle appears to be what's present in the second example as well.

Modifying `Waiter` to use a weak reference to the greenlet doesn't change the behaviour on CPython, however...

**Comment by antocuni** [comment]
> Up until the callback is scheduled, there's no place where the hub would be keeping such a Queue or the consumer greenlet alive. The references to it are contained within the Queue object itself and the stack. And that appears to be exactly what PyPy's GC thinks too: As soon as q.get() is executed and the hub is switched to, the stack becomes non-current and the Queue and greenlet are identified as a sub-graph of objects unreachable from anywhere else and are collected by the producer's gc.collect().

yes exactly, that's what I concluded as well and, and to me it looked like it should be collected even on CPython.

>> Greenlets do not participate in garbage collection; cycles involving data that is present in a greenlet’s frames will not be detected. Storing references to other greenlets cyclically may lead to leaks.

> This would be the cycle: getcurrent() -> `consumer` frame -> Queue -> Waiter -> getcurrent(). The same Waiter is referenced from the stack
> This same cycle appears to be what's present in the second example as well.

I think you nailed it! So, from some point of view it the current implementation works "by chance" on CPython.
Do you think there is a way to change `gevent` to have this case working on PyPy as well?

**Comment by jamadden** [comment]
> Do you think there is a way to change gevent to have this case working on PyPy as well?

Even if there were, and it didn't add any overhead *at all*, I'm not sure that's a good idea. That would be changing "by chance" behaviour into "supported" behaviour. It's a best practice to keep greenlets explicitly alive, IMO, whether keeping a direct reference to them or using APIs that are guaranteed to do so. At least, that's what I always advise my teams to do, largely based on this greenlet documentation:

> If all the references to a greenlet object go away (including the references from the parent attribute of other greenlets), then there is no way to ever switch back to this greenlet. In this case, a GreenletExit exception is generated into the greenlet. 

Of course, it follows up with 

> This feature also enables a programming style in which greenlets are infinite loops waiting for data and processing it. Such loops are automatically interrupted when the last reference to the greenlet goes away.

I don't think the gevent documentation has much to say on this either way. And there may be times when gevent itself relies on that implicitly (I'm thinking of `StreamServer` instances configured without a group or pool. 

Still, I haven't tried yet, but I wonder if a pure-greenlet program written to rely on this behaviour would actually even run the same on CPython and PyPy. Frankly I'm still a bit perplexed that changing the Waiter to use a weak ref didn't change the CPython behaviour in the example. I need to figure out why that is (and that probably involves writing the pure-greenlet test case).

**Comment by antocuni** [comment]
> Even if there were, and it didn't add any overhead at all, I'm not sure that's a good idea. That would be changing "by chance" behaviour into "supported" behaviour. It's a best practice to keep greenlets explicitly alive, IMO, whether keeping a direct reference to them or using APIs that are guaranteed to do so. At least, that's what I always advise my teams to do, largely based on this greenlet documentation:

I agree that it's a poor programming style. My problem is that people try to use their existing program on PyPy, it doesn't work and then blame PyPy :)
Ideally, it would be nice to get a warning if you hit this case on CPython, although I don't see any easy way to do it.
Basically, one of my implicit questions was: does `gevent` guarantee that greenlets created by `gevent.spawn()` are kept alive by `gevent` itself? It seems you are saying that the answer is "no", is it correct?
Would you be ok to add something to `gevent` documentation which explicitly says that you need to keep your greenlets alive, else it's not guaranteed that they will be executed?

> Still, I haven't tried yet, but I wonder if a pure-greenlet program written to rely on this behaviour would actually even run the same on CPython and PyPy. Frankly I'm still a bit perplexed that changing the Waiter to use a weak ref didn't change the CPython behaviour in the example. I need to figure out why that is (and that probably involves writing the pure-greenlet test case).

Good question. I'll try to write such an example and see what I get

**Comment by antocuni** [comment]
> I need to figure out why that is (and that probably involves writing the pure-greenlet test case).

So, you were correct: I managed to write a minimal pure-greenlet test which "fails" on PyPy and does not involve gevent at all:

```python
import sys
import weakref
import gc
import greenlet
print sys.version

class MyQueue(object):
	def get(self):
		self.g = greenlet.getcurrent()
		MAIN.switch()

MAIN = None
G2 = None
WQ = None

def main():
	global MAIN, G2
	MAIN = greenlet.getcurrent()
	G2 = greenlet.greenlet(run_g2)
	switch_to_g2()
	print 'before GC:', WQ()
	gc.collect()
	print 'after GC:', WQ()

def switch_to_g2():
	global G2
	g = G2
	G2 = None
	g.switch()

def run_g2():
	global WQ
	queue = MyQueue()
	WQ = weakref.ref(queue)
	queue.get()

main()
```

```
$ python onlygreenlet.py
2.7.16 (default, Oct  7 2019, 17:36:04) 
[GCC 8.3.0]
before GC: <__main__.MyQueue object at 0x7f60ffeffe50>
after GC: <__main__.MyQueue object at 0x7f60ffeffe50>

$ pypy onlygreenlet.py
2.7.13 (724f1a7d62e8, Dec 23 2019, 15:36:24)
[PyPy 7.3.0 with GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
before GC: <__main__.MyQueue object at 0x00007f70f0fdb718>
after GC: None
```

**Comment by jamadden** [comment]
Thanks, I was looking at that too and came to the same conclusion. 

I think the underlying difference comes down to [this implementation quirk of greenlets](https://github.com/python-greenlet/greenlet/blob/829b44f62b024a3f9b79453b1c0c6a2eb766c336/greenlet.c#L948):

> - frames are not visited: alive greenlets are not garbage collected anyway

Thus, if any function involved in the stack of a running greenlet references (directly or indirectly) the greenlet, there's an uncollectable, unrecognizable, cycle. If there are other strong references to those objects, not a problem. But if there are only weak references outside the cycle, this is visible as an implementation difference. (If there are *no* outside references, it appears to almost be a leak on CPython.)

Another example:

```python
from __future__ import print_function
import sys
import gc
import greenlet

print(sys.version)

MAIN = greenlet.getcurrent()

class MyGreenlet(greenlet.greenlet):
    pass

def find_it():
    gs = [x for x in gc.get_objects() if isinstance(x, MyGreenlet)]
    g = gs[0] if gs else None
    return g


def main():
    global MAIN

    MyGreenlet(run_g2).switch()
    gc.collect()


    print("Main: Greenlet?", find_it())
    gc.collect()
    print("Main: Greenlet refcount?",
          sys.getrefcount(find_it()) if hasattr(sys, 'getrefcount') else -1)
    gc.collect()
    print("Main: Garbage", gc.garbage)

def run_g2():
    print("BG: Running greenlet")
    try:
        print("BG: Find it?", find_it())
        g = greenlet.getcurrent()
        # print("BG: NOT saving getcurrent()")
        MAIN.switch()
    except:
        print("BG: Greenlet being killed", sys.exc_info())
        raise

main()

```

With the assignment of `g = greenlet.getcurrent()`, CPython doesn't collect the greenlet, nor does it kill it as promised, even though the only reference to it is within the stack frame of itself. PyPy cannot find it after a GC cycle.

```console
$ pypy /tmp/g2.py
2.7.13 (724f1a7d62e8, Dec 23 2019, 15:36:47)
[PyPy 7.3.0 with GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)]
BG: Running greenlet
BG: Find it? <__main__.MyGreenlet object at 0x0000000112207700>
BG: NOT saving getcurrent()
Main: Greenlet? None
Main: Greenlet refcount? -1
Main: Garbage []
$ python /tmp/g2.py
3.8.1 (default, Jan 20 2020, 06:58:46)
[Clang 11.0.0 (clang-1100.0.33.17)]
BG: Running greenlet
BG: Find it? <__main__.MyGreenlet object at 0x100cb1f50>
BG: NOT saving getcurrent()
BG: Greenlet being killed (<class 'greenlet.GreenletExit'>, GreenletExit(), <traceback object at 0x100c4caa0>)
Main: Greenlet? None
Main: Greenlet refcount? 4432
Main: Garbage []
```

Comment out the assignment (replace with a print function), and CPython successfully kills the greenlet, as expected:
```console
$ python /tmp/g2.py
3.8.1 (default, Jan 20 2020, 06:58:46)
[Clang 11.0.0 (clang-1100.0.33.17)]
BG: Running greenlet
BG: Find it? <__main__.MyGreenlet object at 0x10f852f50>
BG: NOT saving getcurrent()
BG: Greenlet being killed (<class 'greenlet.GreenletExit'>, GreenletExit(), <traceback object at 0x10f7edb40>)
Main: Greenlet? None
Main: Greenlet refcount? 4438
Main: Garbage []
```

But in what appears to be another implementation difference, PyPy *doesn't kill it* when it garbage collects it.

```console
$ pypy g2.py
2.7.13 (724f1a7d62e8, Dec 23 2019, 15:36:47)
[PyPy 7.3.0 with GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)]
BG: Running greenlet
BG: Find it? <__main__.MyGreenlet object at 0x000000010ad33700>
BG: NOT saving getcurrent()
Main: Greenlet? None
Main: Greenlet refcount? -1
Main: Garbage []
```

**Comment by jamadden** [comment]
The "on-stack references introduce uncollectable cycles" thing in greenlet bugs me, as does the fact that some (hopefully few!) use cases rely on that. It's a bit of implicitness that I don't care for, especially knowing how much work it took us to track down the details of why things are behaving as they are.

Because of:

- the above; and 
- the fact that gevent is supposed to be a higher-level, easier-to-use layer above greenlets; and
- many/most existing gevent blocking APIs already work without keeping greenlet references alive
- the current existence of the notion of a "keep-alive set" that the loops maintain when they go into native code

I think I'm convinced that it would be good to move this from the "by chance" category to the "explicitly designed" category. I think that probably is something like having the `Waiter` record itself in the keep-alive set when it blocks; that seems to solve the issue here.  

I need to study on it a bit more before implementing.

I can't do anything about the lack of a `GreenletKill` exception being raised; that seems like it might be a PyPy bug. gevent itself doesn't rely on that anywhere (it just handles it specially if it happens, but it shouldn't generally [leaving out resurrection] be observable to gevent because by definition the greenlet isn't reachable), but perhaps applications do.

**Comment by antocuni** [comment]
> I think I'm convinced that it would be good to move this from the "by chance" category to the "explicitly designed" category. I think that probably is something like having the Waiter record itself in the keep-alive set when it blocks; that seems to solve the issue here.

I agree: apart from solving the problem on PyPy, this makes gevent more robust against eventual changes in greenlet's behavior.

> I can't do anything about the lack of a GreenletKill exception being raised; that seems like it might be a PyPy bug. gevent itself doesn't rely on that anywhere (it just handles it specially if it happens, but it shouldn't generally [leaving out resurrection] be observable to gevent because by definition the greenlet isn't reachable), but perhaps applications do.

I have opened an issue on the PyPy bug tracker, together with a possible fix, but before committing I am waiting for some feedback from the other PyPy devs: https://bitbucket.org/pypy/pypy/issues/3162/greenletexit-not-thrown-when-a-greenlet-is

**Comment by jamadden** [comment]
The bug report link for [GreenletExit not thrown when a greenlet is garbage-collected](https://foss.heptapod.net/pypy/pypy/issues/3162) has been updated.
